### ISSUE TITLE
Java Grammar: module-info.java support

### ISSUE THREAD

**Issue by pbludov** [checkstyle/checkstyle#8240]
Original issue https://github.com/checkstyle/checkstyle/issues/3059

We should extend our grammar to support module-info.java

- there are may be comments and Javadoc. We should validate them.
- there may be annotations. We have checks for annotations.
- we may add checks specifically for modules. For example, `ModuleName` check.

After completion we should close https://github.com/checkstyle/checkstyle/issues/5816 as it become not required.

**Comment by romani** [comment]
To approve this issue we to clearly decide is it java file or something else.
This will lead us to point of how to fix it:
- update java parser
- create new parser, new Treewalker with file filter(to work only on `module-info.java`)

**Comment by pbludov** [comment]
That's a good point.
Although the JLS explicitly [refers](https://docs.oracle.com/javase/specs/jls/se11/html/jls-7.html#jls-7.7) to the modules as configuration dependency, at the same document it states, that a module name consists of one or more Java identifiers.

On the other hand, it is possible to use "module", "requires", etc as keywords in Java. So it is easier for us to implement a separate grammar and separate checks for modules.
The questionable part is checks for annotations. For example, `AnnotationLocation` check can be applied to modules too. If we can make two grammars that use different numbers for token types, there will be no problems.

**Comment by romani** [comment]
>that a module name consists of one or more Java identifiers.

It is not enough to consider file as java file. C++, C# have a lot in common with Java, but it is not a Java

>The questionable part is checks for annotations. For example, AnnotationLocation check can be applied to modules too

It is better to create new Check for this. In future such config will have more features unrelated to java. Our current Treewalker checks will continue follow java requirements and demands

**Comment by pbludov** [comment]
The list of module-info keywords:
`open, module, requires, transitive, exports, opens, to, uses, provides, with`
[The specs](https://docs.oracle.com/javase/specs/jls/se11/html/jls-3.html#jls-3.9) allows to use them as identifiers in the Java code.

**Comment by Fernal73** [comment]
One rule that I can think of is restrict the use of `open module` only to those modules that are known to be used reflectively. For e.g., module-info.java for test packages that Surefire uses. The rule can either check whether the module definition is under a test directory and/or requires test modules such as those from junit, junit5, testng.

**Comment by nmancus1** [comment]
I would expect something like this as far as the AST structure of a `module-info.java` file:

```
➜  src cat module-info.java
module org.astro {
       exports org.example1 to org.example2;
       requires static org.example2;
}

➜  src java -jar checkstyle-8.44-all.jar -t Test.java

```

```
MODULE_DEF -> MODULE_DEF [1:0]
`--DIRECTIVE_BLOCK -> DIRECTIVE_BLOCK [1:10]
    |--LCURLY -> { [1:10]
    |--EXPORTS -> exports [2:7]
    |   |--PACKAGE_NAME -> org.example1 [2:7]
    |   |--TO -> TO [2:7]
    |   |   `--MODULE_NAME -> org.example2 [2:7]
    |   `--SEMI -> ; [2:16]
    |--REQUIRES -> requires [3:7]
    |   |--MODIFIERS -> MODIFIERS [3:17]
    |   |  `--LITERAL_STATIC -> static [3:17]
    |   |--MODULE_NAME -> org.example2 [3:24]
    |   `--SEMI -> ; [3:36]
    `--RCURLY -> } [4:0]

```

**Comment by joakime** [comment]
What about the ability to checkstyle the module-info canonical order? (like we can with `CustomImportOrder`)

This is the order in use by OpenJDK, and they reference https://docs.oracle.com/javase/specs/jls/se17/html/jls-7.html#jls-ModuleDirective in making that decision.
They enforce it manually (currently) - example - https://github.com/openjdk/jdk/pull/5738

```
module some.module.name {
requires ...

requires transitive ...

exports ...

exports ... to
...

opens ... to
...

uses ...

provides ... with
...
}
```

**Comment by Fernal73** [comment]
It'll be great if it can be applied over entire source.

https://www.javaspecialists.eu/archive/Issue293-Canonical-Orders.html

**Comment by kabutz** [comment]
Submitted a bug report for the jdk.management.jfr module-info.java file, internal bug tracking number 9071602

**Comment by romani** [comment]
I still think it is better to create separate grammar, separate parser ModuleInfoTreeWalker, and make special Check for order

**Comment by pbludov** [comment]
What should be the output of this parser? `DetaiAst` or `DetailNode`? Or, maybe another one?

**Comment by romani** [comment]
New node, it will be like new language support

**Comment by nmancus1** [comment]
> New node, it will be like new language support. Right now we have: java, javadoc. So it will be third language support.

It would be a breaking change, but it would be nice to have consistent naming of AST nodes, if we are going to have different ones. `DetailAST` -> `JavaDetailAst`, `DetailNode` -> `JavadocDetailAst`, and `ModuleInfoDetailAst`.

**Comment by rnveach** [comment]
For a new "TreeWalker" we would need everything under the old one. XPath support and suppression.

> we need to make pre execution filters under Treewalker to avoid module-info.java be parsed by java parser (Treewalker).

Why can't we just hardcode to have TreeWalker ignore `module-info` files and have the new "TreeWalker" only ignore files that aren't `module-info`? Why do users need to configure both TreeWalkers in such a way when the support for parsing is built into both.

**Comment by nmancus1** [comment]
So, to summarize points so far:

- [ ] Implement a new `TreeWalker`, let's call it `ModuleInfoTreeWalker`
- [ ] Add support for XPath, suppressions, etc
- [ ] Create new parser grammar
- [ ] Create new tokens

What else is left to discuss so that we can approve this issue?

**Comment by romani** [comment]
Yes, we can all of this... It is a lot of work.
Let's imagine we have it, how much Checks you can think community need now?
If there a list of best practices for such files ?
If there no much usage...why to do that much work ?

In worse case user can create Fileset Check and parse file by regexp or something simple.

I have never seen such files in open source projects, it might be mostly applicable to projects like jdk, not that much to regular projects.

Hint: we created javadoc grammar only because there were numerous rules and best practices, big documents to describe all rules.

**Comment by joakime** [comment]
@romani the openjdk group has a spec for the order of elements in module-info, and many other projects are following the same spec now.

See: https://docs.oracle.com/javase/specs/jls/se17/html/jls-7.html#jls-ModuleDirective
Mentioned earlier in this issue: https://github.com/checkstyle/checkstyle/issues/8240#issuecomment-931277420

I would like to see a canonical order or import order style of check.
Format / syntax checks (as usual)
optionally a check that ensure a comment exists (above) any use of `requires static` to ensure that a reason is given for them. (several projects I work with have this rule)

**Comment by romani** [comment]
Ok we have 1 potential Check.
I still recommend to not look for first class support now. We can make parser and parse file in each potential Check. While number of Checks is small it will be decently performant and flexible to change at any time it is required to get maturity.
Amount of such files in project is also should not be big, so extra parsing will not be big deal.

First class support will require API change/extension. It is better to avoid it if goal is not big.

**Comment by EM-Creations** [comment]
I'm also encountered this issue at work and it's causing a knock-on effect in SonarQube as it's also using checkStyle..

**Comment by romani** [comment]
@EM-Creations , please use https://checkstyle.org/config_filefilters.html#BeforeExecutionExclusionFileFilter_Examples to avoid problems with such files.

**Comment by EM-Creations** [comment]
> @EM-Creations , please use https://checkstyle.org/config_filefilters.html#BeforeExecutionExclusionFileFilter_Examples to avoid problems with such files.

Hi, that's what I've done but it's not really a fix, it's a work-a-round. The issue still remains in CheckStyle that it doesn't recognise valid Java 9 syntax which was released almost 5 years ago now.

**Comment by romani** [comment]
Wow it is part of Java language https://docs.oracle.com/javase/specs/jls/se9/html/jls-7.html#jls-7.7

**Comment by romani** [comment]
Support of module-info.java is challenging as syntax of this file is not java syntax. But java team did it as it is same by extension, fact that file extension is java, make it complicated for us, due to routing of all java files to Treewalker module, but it knows only java lang syntax.

Workaround now is to create separate checkstyle execution, with special config. In such config, we do have file filter https://checkstyle.org/filefilters/beforeexecutionexclusionfilefilter.html  but we need to use inversion of regexp to match only module files, we have Checker with filter only for java  files, unfortunately Checker can handle  files by file name https://checkstyle.org/config.html#Checker_Properties. In config we have only Regexp like modules and users try to accomplish something by regexp, it is not reliable but better than nothing.

For now, our project is sinking in issues just for regular java files, extension of project scope, distraction from main goal of project, might be not good.

If someone has time, please create project on a side with just another Treewalker for modules to get AST and modules that serve what you need.
One day we will merge such project to common code base to ease usage by users.

**Comment by romani** [comment]
https://github.com/checkstyle/checkstyle/issues/17170#issuecomment-3124742062  and https://github.com/checkstyle/checkstyle/issues/17172#issue-3120033623 describes Google style demands for validation in module-info.java 

demand is growing.....

**Comment by A248** [comment]
If this is not going to be implemented soon, then the least that checkstyle can do is exclude `module-info.java`.

Right now, `module-info.java` can be excluded with the following XML. Until this issue is resolved, the following block should be added by default, so that checkstyle doesn't suddenly break when adding module-info to a project:

```xml
              <module name="BeforeExecutionExclusionFileFilter">
                <property name="fileNamePattern" value="module\-info\.java$"/>
              </module>
```

Not everyone is familiar with checkstyle. So, when you make your tool incompatible with modules by default, you create additional burdens for someone who is trying to migrate existing code and legacy projects to JPMS. Really, checkstyle should not be a stopping block in this, and it's a pain point when tools just refuse to work with JPMS unless you add special configuration.

**Comment by vivek-0509** [comment]
## Summary of this thread

### Problem

Checkstyle's java grammar cannot parse `module-info.java`. Running TreeWalker on a modular compilation unit fails with `IllegalStateException ... no viable alternative at input 'module...'`. The documented workaround is excluding the file via `BeforeExecutionExclusionFileFilter`, which leaves module declarations a permanent blind spot: no AST, no checks, no Javadoc validation. Original report goes back to #3059.

### The two approaches discussed

The central question of this thread was whether module-info is a java file or a separate language:

1. **Extend the java grammar**: modular compilation unit becomes a new top-level alternative of the existing grammar, making sure existing Checks are not affected.
2. **Separate grammar, separate parser, new `ModuleInfoTreeWalker`** dispatched by file name, treating module-info as a third language next to java and javadoc (new node type, new Checks, and a way to keep the java TreeWalker away from module-info files).

Arguments raised for the separate parser: the restricted keywords (`module`, `requires`, ...) are not real java keywords, so a dedicated grammar looks simpler; the java grammar stays untouched; module checks would be free to follow non-java requirements in the future. Open questions it raised: what the output node type would be, whether a new Checker or pre-execution filtering under TreeWalker would be needed, and how checks that apply to both worlds (for example AnnotationLocation) would be shared.

Arguments raised for extending the java grammar: JLS 7.3 defines `CompilationUnit` with `OrdinaryCompilationUnit | CompactCompilationUnit | ModularCompilationUnit` as alternatives of one grammar, and javac parses all three with a single parser; module files reuse most of ordinary Java (lexer, comments and Javadoc, imports, annotations, qualified names), so a unified grammar reuses all of that while a separate parser duplicates it and no existing check works on module files without special wiring.

### Other points collected in the discussion

- The list of restricted keywords: `open, module, requires, transitive, exports, opens, to, uses, provides, with`. All of them stay legal identifiers in ordinary java code (JLS 3.9), so any solution must keep ordinary parsing unchanged.
- A sketch of the expected AST was proposed (MODULE_DEF, DIRECTIVE_BLOCK, directive nodes), with a suggestion to keep qualified names flat so most module checks could be done via MatchXPath. (The implementation follows the sketch's overall shape but keeps qualified names as the usual nested DOT/IDENT trees, consistent with packages and imports, so existing name-based checks work on them.)
- Validation of comments and Javadoc on modules is part of the goal, and annotations on modules should be covered by checks.
- Requested module-specific checks: canonical directive order as OpenJDK enforces manually (#17170), annotation rules for modules (#17172), a ModuleName check, and an idea to restrict `open module` usage.
- After completion, #5816 becomes unnecessary and should be closed.

### Decision

**Approach 1: module support as an extension of the java grammar**. It matches how the JLS and javac structure the language, and the concern that a third top-level branch could repeat the CompactSource problems does not apply here: a module body is a closed set of five directive productions sharing nothing with class bodies, so no existing node type appears in a new position, and the top-level decision happens directly on the `module` keyword, so ambiguity with type declarations is not possible.

**Comment by romani** [comment]
Fix is merged
